### PR TITLE
Converting Crefi to Smallfile changes based #386

### PR DESCRIPTION
**Pull request by AmarnatReddy**
This PR addresses below changes
crefi to smallfile conversion , write_file to remote_file replacements , changing script according to tc description

Logs for [RHEL 8.2 with CEPH 4.2](http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1617362544063/) 
Note : Failures in the logs were because of BZ : [1944943](https://bugzilla.redhat.com/show_bug.cgi?id=1944943)

Please find the logs for execution
[RHEL7.9 with ceph 4.2](http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1618812013243)
This has 4 test cases failed and Test cases have been passed in multiple reruns
[re run log](http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1618906354977/)

[RHEL8.3 with ceph 4.2](http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1618826835422)
This has 3 test cases failed and 1 test case has been passed in [re run](http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1618907283626)

Issue raised for the 2 failed test cases #461

### CODE REVIEW

File: tests/cephfs/CEPH-11219.py [hunk 1]
```diff
@@ -280,7 +283,7 @@ def run(ceph_cluster, **kw):
                     dirs[0],
                     0,
                     1,
-                    iotype="crefi",
+                    iotype="smallfile",
                 )
             print("-------------------------------------------------------")
```
Comment (rakeshgm) [inline tests/cephfs/CEPH-11219.py:288]: can you also remove this `print`

File: tests/cephfs/cephfs_utils.py [hunk 2]
```diff
@@ -29,15 +29,55 @@ def __init__(self, ceph_cluster):
...
+                check_ec=False,
+            )
+
+            # Install pip3
+            client.exec_command(
+                sudo=True,
+                cmd="yum install -y python3-pip",
+                long_running=True,
+                check_ec=False,
+            )
```
Comment (rakeshgm) [inline tests/cephfs/cephfs_utils.py:55]: can you try to combine thes two. lie `packages  = ['python3', 'pip3']` and iterate over.

File: tests/cephfs/cephfs_utils.py [hunk 3]
```diff
@@ -69,40 +109,43 @@ def get_clients(self, build):
...
-                out, rc = node.exec_command(
-                    sudo=True, cmd="pip2  list --format=columns"
-                )
-                output = out.read().decode()
-                output.split()
-                if "crefi" not in output:
-                    node.exec_command(sudo=True, cmd="pip2 install crefi")
+                if "xattr" not in output:
+                    cmds = [
+                        "subscription-manager repos --enable rhel-7-server-optional-rpms "
```
Comment (rakeshgm) [inline tests/cephfs/cephfs_utils.py:125]: we are already enabling `rhel-7-server-optional-rpms`  rpms in install_prereq.py.

File: tests/cephfs/cephfs_utils.py [hunk 4]
```diff
@@ -69,40 +109,43 @@ def get_clients(self, build):
...
+                        "--enable rhel-server-rhscl-7-rpms",
+                        "yum -y install @development",
+                        "yum -y install rh-python36",
+                        "yum -y install rh-python36-numpy rh-python36-scipy",
+                        "yum -y install rh-python36-python-tools rh-python36-python-six",
+                        "yum install -y libffi libffi-devel",
```
Comment (rakeshgm) [inline tests/cephfs/cephfs_utils.py:131]: using a list and iterating makes it easier here too

File: tests/cephfs/cephfs_utils.py [hunk 5]
```diff
@@ -860,12 +910,12 @@ def file_locking(self, clients, mounting_dir):
 try:
     f = open('%sto_test_file_lock', 'w+')
     fcntl.lockf(f, fcntl.LOCK_EX | fcntl.LOCK_NB)
-    print "locking file:--------------------------------"
+    print("locking file:--------------------------------")
```
Comment (rakeshgm) [inline tests/cephfs/cephfs_utils.py:913]: use of `log` makes sense than `print` and the same where ever

File: tests/cephfs/cephfs_utils.py [hunk 6]
```diff
@@ -69,40 +109,43 @@ def get_clients(self, build):
...
-                    sudo=True, cmd="pip2  list --format=columns"
-                )
-                output = out.read().decode()
-                output.split()
-                if "crefi" not in output:
-                    node.exec_command(sudo=True, cmd="pip2 install crefi")
+                if "xattr" not in output:
+                    cmds = [
+                        "subscription-manager repos --enable rhel-7-server-optional-rpms "
+                        "--enable rhel-server-rhscl-7-rpms",
```
Comment (rakeshgm) [inline tests/cephfs/cephfs_utils.py:126]: which rpm would you enable if this being is run on RHEL 8

File: tests/cephfs/cephfs_utils.py (hunk 5 above)
Comment (AmarnatReddy) [reply tests/cephfs/cephfs_utils.py:913]: Hi Rakesh,
print statement was written to a python file.
This python file is copied to the node and we are executing it on the node, so we haven't used log.info

File: tests/cephfs/CEPH-11219.py (hunk 1 above)
Comment (AmarnatReddy) [reply tests/cephfs/CEPH-11219.py]: Hi Rakesh,
I removed all the print statements.

File: tests/cephfs/cephfs_utils.py (hunk 2 above)
Comment (AmarnatReddy) [reply tests/cephfs/cephfs_utils.py]: Hi Rakesh,

I have created a list and iterating through it

File: tests/cephfs/cephfs_utils.py (hunk 4 above)
Comment (AmarnatReddy) [reply tests/cephfs/cephfs_utils.py:121]: Hi Rakesh,

I am iterating and executing the commands

Comment (rakeshgm) [reply tests/cephfs/cephfs_utils.py:121]: i meant, you could use `list` like in the previous file and in the below loop you could use something like this 
```
for pkg in pkgs:
     cmd = f"yum install -y {pkg}"
```     
or all at once 
```
cmd = "yum install -y " + ' '.join(pkgs)
```

File: tests/cephfs/CEPH-10625_11225.py [hunk 7]
```diff
@@ -297,31 +297,16 @@ def run(ceph_cluster, **kw):
...
-                client.exec_command(
-                    cmd="sudo crefi %s'%s' --fop create -t %s "
-                    "--multi -b 10 -d 10 -n 10 -T 10 "
-                    "--random --min=1K --max=%dK"
-                    % (client_info["mounting_dir"], dir_name, "text", 5),
-                    long_running=True,
-                )
-                for i in range(0, 6):
-                    ops = ["create", "rename", "chmod", "chown", "chgrp", "setxattr"]
-                    rand_ops = random.choice(ops)
```
Comment (psathyan) [inline tests/cephfs/CEPH-10625_11225.py:309]: `range` was used to enable random iteration of `ops` however in the new code, this is undone. If the intent was to undone random selection of operations... then you could remove `range` also and iterate through operations.

File: tests/cephfs/CEPH-10625_11225.py [hunk 8]
```diff
@@ -297,31 +297,16 @@ def run(ceph_cluster, **kw):
...
-                        "--multi -b 10 -d 10 -n 10 -T 10 "
-                        "--random --min=1K --max=%dK"
-                        % (
-                            client_info["mounting_dir"],
-                            dir_name,
-                            rand_ops,
-                            rand_filetype,
-                            rand_count,
-                        ),
+                        cmd="sudo python3 smallfile/smallfile_cli.py "
```
Comment (psathyan) [inline tests/cephfs/CEPH-10625_11225.py:303]: Prefer using `client.exec_command(sudo=True, cmd=f"python3 ...")`

File: tests/cephfs/CEPH-11219.py [hunk 9]
```diff
@@ -76,6 +76,9 @@ def run(ceph_cluster, **kw):
         else:
             log.error("Activate multiple mdss failed")
             return 1
+        client1[0].exec_command(
```
Comment (psathyan) [inline tests/cephfs/CEPH-11219.py:79]: Prefer using `sudo=True` and `f-strings` for new code lines.

File: tests/cephfs/CEPH-11222_11223.py [hunk 10]
```diff
@@ -88,10 +88,18 @@ def run(ceph_cluster, **kw):
             )
             log.info("Creating directories with breadth and depth:")
             out, rc = client.exec_command(
-                cmd="sudo crefi %s%s --fop create --multi -b 10 -d 10 "
-                "--random --min=1K --max=10K" % (client_info["mounting_dir"], dir1)
+                cmd="sudo python3 smallfile/smallfile_cli.py "
```
Comment (psathyan) [inline tests/cephfs/CEPH-11222_11223.py:91]: Could we attempt to use `sudo=True`  and `f-strings`

File: tests/cephfs/CEPH-10625_11225.py [hunk 11]
```diff
@@ -297,31 +297,16 @@ def run(ceph_cluster, **kw):
...
-                    long_running=True,
-                )
-                for i in range(0, 6):
-                    ops = ["create", "rename", "chmod", "chown", "chgrp", "setxattr"]
-                    rand_ops = random.choice(ops)
-                    ftypes = ["text", "sparse", "binary", "tar"]
-                    rand_filetype = random.choice(ftypes)
-                    rand_count = random.randint(2, 10)
+                for i in range(0, 5):
+                    ops = ["create", "setxattr", "getxattr", "chmod", "rename"]
```
[the remark this answers is not in the document]
Comment (AmarnatReddy) [reply tests/cephfs/CEPH-10625_11225.py:301]: This was removed as part of yogesh push. It looks  intentional

**Comment by AmarnatReddy** [comment]
I need to execute the cephfs test cases with this latest commits.
I Will update with the logs once RHOS is Up and running

**Comment by bengland2** [comment]
@AmarnatReddy just curious, how's it going with smallfile?

**Comment by AmarnatReddy** [comment]
> @AmarnatReddy just curious, how's it going with smallfile?

@bengland2 I was not able to execute the complete suite as RHOS is down.I will update to you once we execute it.
Hope I understood your question correctly

**Comment by rakeshgm** [comment]
Looks good, have you tried executing this. I think we can executing in RHOS.D now right ?

**Comment by AmarnatReddy** [comment]
> Looks good, have you tried executing this. I think we can executing in RHOS.D now right ?

Hi Rakesh,
As **cislave.ceph.redhat.com** is down. I am running the tests from my laptop so it is taking long time.
I am executing it on RHEL 8 and RHEL 7

**Review by psathyan — APPROVED** [review]

**Review by rakeshgm — APPROVED** [review]

**Pull request merged**